internal/controller/argocdcommitstatus_controller.go

```diff
@@ -0,0 +1,398 @@
+			phase = promoterv1alpha1.CommitPhaseFailure
+		}
+
+		// This is an in memory version of the desired CommitStatus for a single application, this will be used to figure out
+		// the aggregated phase of all applications for a particular environment
+		aggregateItem.commitStatus = &promoterv1alpha1.CommitStatus{
+			Spec: promoterv1alpha1.CommitStatusSpec{
+				Sha:   application.Status.Sync.Revision,
+				Phase: phase,
+			},
+		}
+		aggregateItem.selectedApplication = &promoterv1alpha1.ApplicationsSelected{}
+		aggregateItem.selectedApplication.Phase = phase
+		aggregateItem.selectedApplication.Sha = application.Status.Sync.Revision
+		aggregateItem.selectedApplication.Name = application.GetName()
+		aggregateItem.selectedApplication.Namespace = application.GetNamespace()
+		aggregateItem.selectedApplication.LastTransitionTime = application.Status.Health.LastTransitionTime
+
+		argoCDCommitStatus.Status.ApplicationsSelected = append(argoCDCommitStatus.Status.ApplicationsSelected, promoterv1alpha1.ApplicationsSelected{
+			Namespace:          application.GetNamespace(),
+			Name:               application.GetName(),
+			Phase:              phase,
+			Sha:                application.Status.Sync.Revision,
+			LastTransitionTime: application.Status.Health.LastTransitionTime,
+		})
+
+		aggregates[key] = append(aggregates[key], aggregateItem)
+	}
+
+	for key, aggregateItem := range aggregates {
+		var resolvedSha string
+		repo, targetBranch := key.repo, key.targetBranch
+
+		gitOperation, err := git.NewGitOperations(ctx, r.Client, gitAuthProvider, r.PathLookup, repositoryRef, &argoCDCommitStatus, targetBranch)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to initialize git client: %w", err)
+		}
+
+		resolvedSha, err = gitOperation.LsRemote(ctx, targetBranch)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to ls-remote sha: %w", err)
+		}
+
+		var desc string
+		resolvedPhase := promoterv1alpha1.CommitPhasePending
+		pending := 0
+		healthy := 0
+		degraded := 0
+		var mostRecentLastTransitionTime *metav1.Time
+		for _, s := range aggregateItem {
+			if s.commitStatus.Spec.Sha != resolvedSha {
+				pending++
+			} else if s.commitStatus.Spec.Phase == promoterv1alpha1.CommitPhaseSuccess {
+				healthy++
+			} else if s.commitStatus.Spec.Phase == promoterv1alpha1.CommitPhaseFailure {
+				degraded++
+			} else {
+				pending++
+			}
+
+			// Find the most recent last transition time
+			if s.selectedApplication.LastTransitionTime != nil {
+				if mostRecentLastTransitionTime == nil || s.selectedApplication.LastTransitionTime.Time.After(mostRecentLastTransitionTime.Time) {
+					mostRecentLastTransitionTime = s.selectedApplication.LastTransitionTime
+				}
+			}
+		}
+
+		// Did the mostRecentLastTransitionTime occur more than 5 seconds ago
+		if mostRecentLastTransitionTime != nil && time.Since(mostRecentLastTransitionTime.Time) < 5*time.Second {
```

Comment:
I'm a little worried about this check. Seems like if we have one flappy app, we never update the CommitStatus. Maybe we should update everything except the Phase?